chiselFrontend/src/main/scala/chisel3/NumBits.scala

```diff
@@ -0,0 +1,140 @@
+   * @group Bitwise
+   */
+  def << (that: Int): T
+
+  /** @group SourceInfoTransformMacro */
+  def do_<< (that: Int)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T
+
+  /** Dynamic left shift operator
+   *
+   * @param that a hardware component
+   * @return this $coll dynamically shifted left by `that` many places, shifting in zeros from the right
+   * @note The width of the returned $coll is `width of this + pow(2, width of that) - 1`.
+   * @group Bitwise
+   */
+  def << (that: UInt): T
+
+  /** @group SourceInfoTransformMacro */
+  def do_<< (that: UInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T
+
+  /** Static right shift operator
+   *
+   * @param that an amount to shift by
+   * @return this $coll with `that` many least significant bits truncated
+   * $unchangedWidth
+   * @group Bitwise
+   */
+  // REVIEW TODO: redundant
+  def >> (that: BigInt): T
+
+  /** @group SourceInfoTransformMacro */
+  def do_>> (that: BigInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T
+
+  /** Static right shift operator
+   *
+   * @param that an amount to shift by
+   * @return this $coll with `that` many least significant bits truncated
+   * $unchangedWidth
+   * @group Bitwise
+   */
+  def >> (that: Int): T
+
+  /** @group SourceInfoTransformMacro */
+  def do_>> (that: Int)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T
+
+  /** Dynamic right shift operator
+   *
+   * @param that a hardware component
+   * @return this $coll dynamically shifted right by the value of `that` component, inserting zeros into the most
+   * significant bits.
+   * $unchangedWidth
+   * @group Bitwise
+   */
+  def >> (that: UInt): T
+
+  /** @group SourceInfoTransformMacro */
+  def do_>> (that: UInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T
+
+  // TODO: refactor to share documentation with Num or add independent scaladoc
+  /** Unary negation (expanding width)
+   *
+   * @return a $coll equal to zero minus this $coll
+   * $constantWidth
+   * @group Arithmetic
+   */
+  final def unary_- (): T = macro SourceInfoTransform.noArg
+
+  /** Unary negation (constant width)
+   *
+   * @return a $coll equal to zero minus this $coll shifted right by one.
+   * $constantWidth
+   * @group Arithmetic
+   */
+  final def unary_-% (): T = macro SourceInfoTransform.noArg
+
+  override def do_< (that: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Bool =
+    compop(sourceInfo, LessOp, that)
+  override def do_> (that: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Bool =
+    compop(sourceInfo, GreaterOp, that)
+  override def do_<= (that: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Bool =
+    compop(sourceInfo, LessEqOp, that)
+  override def do_>= (that: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Bool =
+    compop(sourceInfo, GreaterEqOp, that)
```

Comment:
Move these operators into `Num`.

Reply:
It won't be possible to move the implementations to `Num` as it does not require the sub-classes to extend `Bits`, and the `compop` is defined in `Bits`. I was also under the impression that `Num` is not supposed to have specific implementation as it's part of the user API. Is there currently consensus on what `Num` actually is?

Reply:
Unfortunately, there's no good consensus. It's something that we need to figure out... If you have ideas, it would be reasonable to propose something.